### PR TITLE
Add ContiguousDisjointSet datastructure

### PR DESCRIPTION
**Pull request by clonker**
needed for loop nesting forest construction in liveness analysis

### CODE REVIEW

File: libsolutil/DisjointSet.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	return rootElement;
+}
+
+bool ContiguousDisjointSet::merge(value_type const _x, value_type const _y, bool const _mergeBySize)
+{
+	auto xRoot = find(_x);
+	auto yRoot = find(_y);
+
+	if (xRoot == yRoot)
+		return false;  // we're done, nothing to merge here
```
Comment (nikola-matic) [inline libsolutil/DisjointSet.cpp]: Is this actually useful for anything other than testing? I would assume if you want to merge two sets, all you ultimately care about is for the union operation to produce - well, a union of two sets; whether the `_x` and `_y` were already members of the same set seems like irrelevant information here.

**Review by nikola-matic — COMMENTED** [review]
Is this for SSA CFG?

**Comment by clonker** [comment]
> Is this for SSA CFG?

Yeah, this is a data structure used in finding / construction of the loop nesting forest of a CFG

File: test/libsolutil/DisjointSet.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Unit tests for FixedHash.
```
Comment (r0qs) [inline test/libsolutil/DisjointSet.cpp]:
```suggestion
 * Unit tests for DisjointSet.
```

File: libsolutil/DisjointSet.cpp (hunk 1 above)
Comment (r0qs) [reply libsolutil/DisjointSet.cpp]: I believe this is done because the root nodes in this data structure represent the sets they contain, so two nodes are in the same set if and only if the roots of the trees containing the nodes are equal. Thus it is not necessary to merge in such case.

Comment (clonker) [reply libsolutil/DisjointSet.cpp]: It is not strictly required and indeed, for now I just used it for testing. At first I was thinking it could give you some performance benefit to return the status code here but in reality this isn't even true - let's remove it:)

Comment (nikola-matic) [reply libsolutil/DisjointSet.cpp]: Yup, I know that - I'm just arguing that the `merge` function (which should be called `union`) doesn't need to return anything, i.e. it should be `void`. I'm simply asking whether having it returning a boolean indicating that a union operation was performed (or whether the two elements were already in the same set) is going to be relevant information when used within it the liveness analysis?

Comment (clonker) [reply libsolutil/DisjointSet.cpp]: I would have called it union if union wasn't a reserved keyword in c++ ;)

Comment (nikola-matic) [reply libsolutil/DisjointSet.cpp]: `uNiOn`. I personally prefer merge, because it's more comp sci and less math :D

Comment (clonker) [reply libsolutil/DisjointSet.cpp]: boo! :D i kind of like your suggestion, has that sarcastic aftertaste.. let's make a `uNi0n` of these two sets!

File: libsolutil/DisjointSet.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	value_type rootElement = _element;
+	while (rootElement != m_parents[rootElement])
+	{
+		m_parents[rootElement] = m_parents[m_parents[rootElement]];
+		rootElement = m_parents[rootElement];
+	}
+	return rootElement;
+}
+
+void ContiguousDisjointSet::merge(value_type const _x, value_type const _y, bool const _mergeBySize)
```
Comment (matheusaaguiar) [inline libsolutil/DisjointSet.cpp:53]: Are we planning to add rank in the future? Or is there a situation where we would have this false?
just a nitpick...

ps: I see it set to false in the test, but realistically, should be better to always have it on (or switched to rank), right?

**Review by matheusaaguiar — APPROVED** [review]
I checked the algorithm and can't see anything wrong.

File: libsolutil/DisjointSet.cpp (hunk 3 above)
Comment (clonker) [reply libsolutil/DisjointSet.cpp:53]: Good question:) The place we need it for now (loop detection with Tarjan's algorithm) relies on merging with "by size/rank" being switched off. That way a representative of a set can stay stable under merge operations. In particular this is used for loop headers (think a loop header as a vertex of a CFG represent a set of vertices that comprise the loop body). You can see this in [PR 15419](https://github.com/ethereum/solidity/pull/15419).

So generally I am inclined to agree that by rank is better - although potential problems with "by-size" are somewhat mitigated by path compression - but in our case it doesn't really matter. For now. :D

Comment (clonker) [reply libsolutil/DisjointSet.cpp:53]: Oh and there is just one test that has `mergeBySize` disabled, to check exactly that stability property of the set representative... the rest does have it enabled..

Comment (matheusaaguiar) [reply libsolutil/DisjointSet.cpp:53]: Thanks for the explanation @clonker !
